src/data/pnat/find.lean

```diff
@@ -0,0 +1,133 @@
+/-
+Copyright (c) 2022 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky, Floris van Doorn
+-/
+import data.pnat.basic
+
+/-!
+# Explicit least witnesses to existentials on positive natural numbers
+
+Implemented via calling out to `nat.find`.
+
+-/
+
+namespace pnat
+
+variables {p q : ℕ+ → Prop} [decidable_pred p] [decidable_pred q] (h : ∃ n, p n)
+
+instance decidable_pred_exists_nat :
+  decidable_pred (λ n' : ℕ, ∃ (n : ℕ+) (hn : n' = n), p n) := λ n',
+decidable_of_iff' (∃ (h : 0 < n'), p ⟨n', h⟩) $ subtype.exists.trans $
+  by simp_rw [subtype.coe_mk, @exists_comm (_ < _) (_ = _), exists_prop, exists_eq_left']
+
+
+include h
+
+/--
+If `p` is a (decidable) predicate on `ℕ+` and `hp : ∃ (n : ℕ+), p n` is a proof that
+there exists some positive natural number satisfying `p`, then `pnat.find hp` is the
+smallest positive natural number satisfying `p`. Note that `pnat.find` is protected,
+meaning that you can't just write `find`, even if the `pnat` namespace is open.
+
+The API for `pnat.find` is:
+
+* `pnat.find_spec` is the proof that `pnat.find hp` satisfies `p`.
+* `pnat.find_min` is the proof that if `m < pnat.find hp` then `m` does not satisfy `p`.
+* `pnat.find_min'` is the proof that if `m` does satisfy `p` then `pnat.find hp ≤ m`.
+-/
+protected def find_aux
+  (H' : ∃ (n' : ℕ) (n : ℕ+) (hn' : n' = n), p n := exists.elim h (λ n hn, ⟨n, n, rfl, hn⟩)) : ℕ :=
+nat.find H'
+
+protected theorem find_aux_spec
+  (H' : ∃ (n' : ℕ) (n : ℕ+) (hn' : n' = n), p n := exists.elim h (λ n hn, ⟨n, n, rfl, hn⟩)) :
+  ∃ (n : ℕ+) (hn' : pnat.find_aux h = n), p n :=
+nat.find_spec H'
```

Comment:
Is it easier to go straight to `nat.find_x`?

Reply:
Can you clarify? Do you mean using `nat.find_x` directly in the definition of `pnat.find`? I would still need to construct the nat-existential term, and I am trying to launder it through these aux functions.

Reply:
I'm suggesting you launder it through a single `find_x_aux` function rather than needing three `aux` functions

Reply:
I have only one `aux` function, `pnat.find_aux`. I'm sorry, I don't understand your suggestion.